### PR TITLE
Add metrics exchange mechanism

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
### Description

We create this PR to support metrics exchange (or logging any text, scalar, scalars)
`MetricRelay` <-> `FlareAgent` has similar relationship as `TaskExchanger` <-> `FlareAgent`.

Adds:

- `MetricRelay`, it will get log values (text, scalar, scalars) from external and then fire a event
- `ExternalConfigurator`, it will prepare the config needed, it will go through the components list and call `AttributesExportable`'s `export` method
- An example of "np_metrics" showcase how to send log values back (with corresponding job_template)
- A ClientAPI `flare.log`
- High level APIs nvflare.client.tracking: `SummaryWriter`, `WandBWriter`, `MLflowWriter`

Updates:

- examples/hello-world/ml-to-fl/pt/code/cifar10_fl.py to showcase how to log values
- move send_analytic_dxo and create_analytic_dxo from nvflare/app_common/widgets/streaming.py to nvflare/apis/utils/analytix_utils.py as they are general and can be used by other components
- `MetricsRetriever` to be based on `MetricRelay`
- `MemoryPipe` to be configurable in config

Removes:
- `Piper`, use `ExternalConfigurator` instead

TODO in follow up PRs:
- api should not depend on app_common: from nvflare.app_common.tracking.tracker_types import LogWriterName, TrackConst
- This changes will touch multiple places so defer to the next PR

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: job_templates/sag_np_metrics/config_fed_client.conf [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      id = "metric_relayer"
+      path = "nvflare.app_common.metrics_exchange.client_api_metric_relayer.ClientAPIMetricRelayer"
+      args {
+        pipe_id = "metrics_pipe"
+        event_type = "analytix_log_stats"
+        # how fast should it read from the peer
+        read_interval = 0.1
+      }
+    },
+    {
```
Comment (ZiyueXu77) [inline job_templates/sag_np_metrics/config_fed_client.conf:110]: shall we hide / combine this with relayer

File: job_templates/sag_np_metrics/config_fed_client.conf [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      id = "event_to_fed"
+      path = "nvflare.app_common.widgets.convert_to_fed_event.ConvertToFedEvent"
+      args {
+        events_to_convert = ["analytix_log_stats"]
+        fed_event_prefix = "fed."
```
Comment (ZiyueXu77) [inline job_templates/sag_np_metrics/config_fed_client.conf:115]: shall we hide event_to_fed or combine this with relayer?

File: nvflare/app_common/metrics_exchange/client_api_metric_relayer.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.data_exchange.piper import Piper
+from nvflare.client.config import ClientConfig, ConfigKey
+from nvflare.client.constants import CONFIG_METRICS_EXCHANGE
+
+from .metric_relayer import MetricRelayer
+
+
+class ClientAPIMetricRelayer(MetricRelayer):
```
Comment (YuanTingHsieh) [inline nvflare/app_common/metrics_exchange/client_api_metric_relayer.py:26]: change to `ClientAPIConfigPreparer`

File: nvflare/app_common/metrics_exchange/memory_metric_relayer.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.metrics_exchange.metrics_exchanger import MetricsExchanger
+from nvflare.fuel.utils.constants import Mode
+from nvflare.fuel.utils.pipe.memory_pipe import MemoryPipe
+from nvflare.fuel.utils.pipe.pipe_handler import PipeHandler
+
+from .metric_relayer import MetricRelayer
+
+
+class MemoryMetricRelayer(MetricRelayer):
```
Comment (YuanTingHsieh) [inline nvflare/app_common/metrics_exchange/memory_metric_relayer.py:27]: keep old name "MetricsRetriever",

Add note it only exists for backward compatibility

File: examples/hello-world/ml-to-fl/np/README.md [hunk 5]
```diff
@@ -143,3 +143,28 @@ Then we can run it using the NVFlare Simulator:
 ```bash
 nvflare simulator -n 2 -t 2 ./jobs/np_loop_cell_pipe -w np_loop_cell_pipe_workspace
 ```
+
+## Launch once for the whole job and with metrics streaming
+
+Sometimes we want to stream the training progress.
+
+We can use the flare.log method to do that: [./code/train_metrics.py](./code/train_metrics.py)
```
Comment (holgerroth) [inline examples/hello-world/ml-to-fl/np/README.md:151]: I don't see flare.log being used in train_metrics.py.

Comment (YuanTingHsieh) [reply examples/hello-world/ml-to-fl/np/README.md]: fixed

File: examples/hello-world/ml-to-fl/pt/code/cifar10_fl.py [hunk 6]
```diff
@@ -22,6 +22,9 @@
 # (1) import nvflare client API
 import nvflare.client as flare
 
+# (optional) metrics
+from nvflare.client.tracking import TBWriter
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply examples/hello-world/ml-to-fl/pt/code/cifar10_fl.py]: It does NOT depend on server to use what framework though, users can use `SummaryWriter` here and server side configure `MLflowReceiver`.
Or `MLflowWriter` and server side use `TBReceiver` (which is the case for example np_metrics)

File: nvflare/app_common/widgets/client_api_configurator.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        components = engine.get_all_components()
+        components_data = {}
+        for component_id in self._component_ids:
+            component_instance: ClientAPIExportable = components.get(component_id)
+            if component_instance is None:
+                raise RuntimeError(f"missing component: {component_id}")
+            component_attributes = component_instance.export_for_client_api()
+            for k, v in component_attributes.items():
+                if k in components_data:
+                    raise RuntimeError(
```
Comment (yanchengnv) [inline nvflare/app_common/widgets/client_api_configurator.py:116]: I don't think we can require all components to have globally unique attributes.
The config structure should have separate sections for different components. It should be okay for different components to use the same attr names.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Please see my comments.

File: nvflare/fuel/utils/pipe/memory_pipe.py [hunk 8]
```diff
@@ -20,23 +20,34 @@
...
 
-    def open(self, name: str):
-        pass
-
     def clear(self):
         pass
 
     def close(self):
-        pass
+        if self.token in MemoryPipe.PIPE_PAIRS:
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/fuel/utils/pipe/memory_pipe.py]: thanks, fixed as suggested

File: nvflare/client/tracking.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+from typing import Dict, Optional
+
+from nvflare.apis.analytix import AnalyticsDataType
+from nvflare.app_common.tracking.tracker_types import LogWriterName
+
+from .api import log
+
+
+class SummaryWriter:
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/client/tracking.py:23]: yes, we have similar interfaces here and there, follow up PR we can clean/re-factor implementation in MONAI integration side.

File: nvflare/app_common/widgets/client_api_configurator.py (hunk 7 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/widgets/client_api_configurator.py]: thanks, fixed as suggested

File: nvflare/fuel/utils/attributes_exportable.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    @abstractmethod
+    def export(self, export_mode: str) -> Tuple[str, dict]:
+        """Exports attributes.
+
+        Args:
+            export_mode (str): export to peer (`ExportMode.PEER`) or to self (`ExportMode.SELF`).
+
+        Returns:
+            A tuple of (export section name, arguments to be exported)
```
Comment (yanchengnv) [inline nvflare/fuel/utils/attributes_exportable.py:37]: I see that you sometimes return class name (for pipe) and sometimes return a section name (e.g. metrics). We should have a consistent definition. We can improve this later.

**Pull request merged**